### PR TITLE
[NETBEANS-1965] Export public & friend packages for Payara Plugin

### PR DESCRIPTION
**Pull request by jGauravGupta**

### CODE REVIEW

**Comment by geertjanw** [comment]
Why does org.netbeans.modules.openide.windows need to be public?

**Comment by jGauravGupta** [comment]
Payara Micro plugin's  context service provider supersedes the `org.netbeans.modules.openide.windows.GlobalActionContextImpl`.

**Comment by geertjanw** [comment]
Only for Windows, not for other operating systems? And I suppose you're using 'supersedes' in the annotation and you need to expose this module because you need to depend on it for ordering purposes? It sounds a little bit suspicious.

**Comment by geertjanw** [comment]
Essentially, while I agree completely with the aim behind this PR, we're creating a bunch of new APIs here -- what's the process we should follow for that?

**Comment by geertjanw** [comment]
But won’t that supersede the default context provider and thereby change the context completely throughout the IDE for all features of the IDE?

**Comment by geertjanw** [comment]
A different approach to consider would be to donate the Payara plugin source code to Apache NetBeans, i.e., to have it in the trunk of Apache NetBeans and to develop it there. It would be a fantastic contribution to NetBeans, especially since the next release, targeted to be released in March, will focus on the enterprise, i.e., Jakarta EE cluster.

**Comment by jGauravGupta** [comment]
Exactly, we (Payara Team members) already had discussion about it last week.

**Comment by geertjanw** [comment]
That would mean you would need to make none of (or less of) these friend requests and you would not need to expose any (or several) of these APIs. Instead, you'd work with the Payara source code as a standard part of Apache NetBeans.

**Comment by jGauravGupta** [comment]
Regarding Payara plugin donation, we had just discussion, so officially not approved yet. I hope so once approved by both party, It will start after next Payara Server release.

> That would mean you would need to make none of (or less of) these friend requests

I understand that most of APIs don't need to be expose for internal modules. But can't find much difference for friend request, As friend requests are also added for GlassFish modules.

**Comment by geertjanw** [comment]
Yes, that's true, and you're right. Looking forward to this a lot.

**Comment by jlahoda** [comment]
-1 on the current patch, sorry.

I think adding an (external) module as a friend to modules that already have an external friend is not very controversial.

What is controversial is opening new packages. I think that in all cases, it is necessary to start with a description why a particular package is opened.

The most controversial are cases where a package is opened publicly. Some cases (like opening org.netbeans.modules.openide.windows), seem to be simply undesirable, and any usecase which needed that should be implemented as a proper API enhancement.

Specifically on the org.netbeans.modules.openide.windows package: overriding GlobalActionContextImpl will affect the whole IDE, I believe, so some care needs to be taken. But I don't see why the package must be public. ServiceProvider allows to supersede (mask out) other implementation given their name, no need to make the package public. So what is exactly the usecase?

Thanks.

File: enterprise/j2ee.sun.dd/nbproject/project.xml [hunk 1]
```diff
@@ -141,8 +143,48 @@
                 <package>org.netbeans.modules.j2ee.sun.dd.api.serverresources</package>
                 <package>org.netbeans.modules.j2ee.sun.dd.api.services</package>
                 <package>org.netbeans.modules.j2ee.sun.dd.api.web</package>
+                <package>org.netbeans.modules.j2ee.sun.dd.impl</package>
```
Comment (JaroslavTulach) [inline enterprise/j2ee.sun.dd/nbproject/project.xml]: -1 on these. There is a reason it is called `impl`.

**Comment by JaroslavTulach** [comment]
-10

**Pull request closed without merging**